plugins/headless/hu.bme.mit.gamma.api.headless/src/hu/bme/mit/gamma/api/headless/Application.java

```diff
@@ -0,0 +1,186 @@
+        ActionLanguageStandaloneSetup.doSetup();
+        StatechartLanguageStandaloneSetup.doSetup();
+        TraceLanguageStandaloneSetup.doSetup();
+        PropertyLanguageStandaloneSetup.doSetup();        
+    	GenModelStandaloneSetup.doSetup();
+        final Map<?, ?> args = context.getArguments();
+        final String[] appArgs = (String[]) args.get(IApplicationContext.APPLICATION_ARGS);
+        if (appArgs.length >= 1) {
+            String ggenFilePath = URI.decode(appArgs[0]);
+            File ggenFile = new File(ggenFilePath);
+            String projectDescriptorPath = URI.decode(appArgs[1]);
+            File projectFolder = getContainingProject(ggenFile);
+            String projectName = projectFolder.getName();
+            String fileWorkspaceRelativePath = ggenFilePath.substring(projectFolder.getParent().length());
+            IWorkspace workspace = ResourcesPlugin.getWorkspace();
+            IWorkspaceRoot workspaceRoot = workspace.getRoot();
+            IPath workspaceLocation = workspaceRoot.getLocation();
+            File workspaceFolder = workspaceLocation.toFile();
+            IProgressMonitor progressMonitor = new NullProgressMonitor();
+            // The file and its containing project is not in the given workspace
+            // The project has to be copied into the workspace
+            if (!contains(workspaceFolder, ggenFile)) {
+                // Note that in this case the ggen cannot refer to models outside of the project
+                IProject project = workspaceRoot.getProject(projectName);
+                try {
+                    project.create(progressMonitor);
+                } catch (CoreException creationException) {
+                    // Project with same name exists, trying to open it
+                    try {
+                        project.open(progressMonitor);
+                    } catch (CoreException openException) {
+                        // Open did not succeed, deleting and creating needed
+                        project.delete(true, progressMonitor);
+                        project.create(progressMonitor);
+                    }
+                }
+                project.open(progressMonitor);
+                //
+                IProjectDescription description = project.getDescription();
+//              description.setNatureIds(new String[] { XtextProjectHelper.NATURE_ID });
+                project.setDescription(description, progressMonitor);
+                // Not needed to add project natures like this, maybe copyDirectory does that?
+                copyDirectory(projectFolder, project);
+                workspace.save(true, progressMonitor);
+            }
+            // The file and its containing project is in the given workspace
+            GammaApi gammaApi = new GammaApi();
+            gammaApi.run(fileWorkspaceRelativePath, new ResourceSetCreator() {
+				public ResourceSet createResourceSet() {

```

Comment:
I think the Injector does not have to be retrieved every time when we create a ResourceSet. Let us try to introduce a "singleton" pattern for the injector:
```
private Injector injector = null;
private Injector getInjector() {
	if (injector == null) {
 		injector = new  StatechartLanguageStandaloneSetupGenerated().createInjectorAndDoEMFRegistration();						
	}
	return injector;
}
public ResourceSet getResourceSet() {
	Injector injector = getInjector();
	XtextResourceSet resourceSet = injector.getInstance(XtextResourceSet.class);
	return resourceSet;
}
```